app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataCache.kt

```diff
@@ -0,0 +1,491 @@
+
+            for (relation in updatedRelations) {
+                // old relation may now have different members, so they need to be removed first
+                val oldRelation = relationCache[relation.id]
+                if (oldRelation != null) {
+                    for (oldMember in oldRelation.members) {
+                        val memberKey = ElementKey(oldMember.type, oldMember.ref)
+                        relationIdsByElementKeyCache[memberKey]?.remove(relation.id)
+                    }
+                }
+                relationCache[relation.id] = relation
+
+                /// ...and then the new members added
+                for (member in relation.members) {
+                    val memberKey = ElementKey(member.type, member.ref)
+                    // only if the node member is already in the spatial cache or any node of a member
+                    // is, the relation ids it refers to must be known:
+                    // TODO explain why
+                    val isInSpatialCache = when (member.type) {
+                        ElementType.NODE -> spatialCache.get(member.ref) != null
+                        ElementType.WAY -> member.ref in wayIds
+                        ElementType.RELATION -> member.ref in relationIds
+                    }
+                    val relationIdsReferredByMember = if (isInSpatialCache) {
+                        relationIdsByElementKeyCache.getOrPut(memberKey) { ArrayList(2) }
+                    } else {
+                        relationIdsByElementKeyCache[memberKey]
+                    }
+                    relationIdsReferredByMember?.add(relation.id)
+                }
+            }
+        }
+    }
+
+    /**
+     * Gets the element with the given [type] and [id] from cache. If the element is not cached,
+     * [fetch] is called, and the result is cached and then returned.
+     */
+    fun getElement(
+        type: ElementType,
+        id: Long,
+        fetch: (ElementType, Long) -> Element?
+    ): Element? = synchronized(this) {
+        val element = when (type) {
+            ElementType.NODE -> spatialCache.get(id)
+            ElementType.WAY -> wayCache.getOrPutIfNotNull(id) { fetch(type, id) as? Way }
+            ElementType.RELATION -> relationCache.getOrPutIfNotNull(id) { fetch(type, id) as? Relation }
+        }
+        return element ?: fetch(type, id)
+    }
+
+    /**
+     * Gets the geometry of the element with the given [type] and [id] from cache. If the geometry
+     * is not cached, [fetch] is called, and the result is cached and then returned.
+     */
+    fun getGeometry(
+        type: ElementType,
+        id: Long,
+        fetch: (ElementType, Long) -> ElementGeometry?
+    ): ElementGeometry? = synchronized(this)  {
+        val geometry = when (type) {
+            ElementType.NODE -> spatialCache.get(id)?.let { ElementPointGeometry(it.position) }
+            ElementType.WAY -> wayGeometryCache.getOrPutIfNotNull(id) { fetch(type, id) }
```

Comment:
Here and elsewhere where `getOrPutIfNotNull` is used:

Why put it only if it is not null? And why, then, is `wayGeometryCache` a HashMap<Long, ElementGeometry<b><i>?</i></b>> (emphasis on the `? `) if it is only ever filled if the value is non-null?

Reply:
I already mentioned this in some other comment above. I am not 100% sure, but think this has historic reasons, when at some point the cached elements could not be `null`

Reply:
Switched to `getOrPutNullable` in https://github.com/streetcomplete/StreetComplete/pull/4125/commits/985ef48a703c9092a6323152774d6d3cac435302
And added caching `null` for missing keys in `getElements`/`getGeometries` in https://github.com/streetcomplete/StreetComplete/pull/4125/commits/e00c3ee67db84cd968eaf83f2815ae87fea8843a

Reply:
I've looked at https://github.com/streetcomplete/StreetComplete/pull/4125/commits/985ef48a703c9092a6323152774d6d3cac435302 and https://github.com/streetcomplete/StreetComplete/pull/4125/commits/e00c3ee67db84cd968eaf83f2815ae87fea8843a and the implementation looks really effortful, first with checking whether it's `null` but already cached, then looping to the `keysToFetch` to find the difference between those that were returned from db and those that weren't, to put them into cache but as `null` after all.

So, makes me wonder if this is necessary. 

When are these method called and when can a queried element be `null`? Ultimately:
1. `getGeometries` is called through `OsmQuestController.getAllVisibleInBBox` and `OsmQuestController.getAllHiddenNewerThan`. But only for elements the quests refer to, so there should be zero chance that any of those geometries is not in the db. Same with `getGeometry`
2. `getWays` is called through `MapDataController.completeMapData`. There it can actually frequently happen that some ways that are members of a relations are not in the db. `completeMapData` is called after download bbox and upload edit and they contained (incomplete) relations. Which relations are most often incomplete and contain vast numbers of ways? Route relations. Really not sure about other relations.

That's what I could find. So having queries for specific node, way or relation ids return null seems to be rather the exceptional case, hence not worth optimizing for.

Reply:
Oh, I didn't even notice this comment... probably because there's only a single notification for all updates in the PR.

I just added some logging to see whether storing null ever happens, but in a (rather sshort/simple) test didn't find anything.

Reply:
In a longer test logging triggered once for `getElements`, i.e. once it `getElements` avoided a `fetch` because of storing null.
Don't know what exactly caused it, but once in ca 1.5 hours / 400 edits is probably nothing to worry about.

So... simply remove the capability of storing `null` values?

Reply:
👍 

It would happen if a relation is downloaded only incompletely. Maybe a complex building with holes where one hole was outside the bbox.